app/Models/Concerns/HasRoles.php

```diff
@@ -0,0 +1,277 @@
+    }
+
+    /**
+     * Scope the model query to certain roles only.
+     *
+     * @param Builder $query
+     * @param string|array|Collection $roles
+     * @return Builder
+     */
+    public function scopeRole(Builder $query, $roles): Builder
+    {
+        if ($roles instanceof Collection) {
+            $roles = $roles->all();
+        }
+        if (! is_array($roles)) {
+            $roles = [$roles];
+        }
+        $roles = array_map(function ($role) {
+            if ($role instanceof Role) {
+                return $role;
+            }
+            $method = is_numeric($role) ? 'find' : 'findByName';
+
+            return Role::{$method}($role);
+        }, $roles);
+
+        return $query->whereHas('roles', function ($query) use ($roles) {
+            $query->where(function ($query) use ($roles) {
+                foreach ($roles as $role) {
+                    $query->orWhere('roles.id', $role->id);
+                }
+            });
+        });
+    }
+
+    /**
+     * Assign the given role to the model.
+     *
+     * @param Collection|array|string ...$roles
+     *
+     * @return $this
+     */
+    public function assignRole(...$roles): self
+    {
+        $roles = $this->roleInputToIds($roles)->all();
+
+        $changes = $this->roles()->sync($roles, false);
+
+        if ($this instanceof User && collect($changes)->sum(function ($value) {
+            return count($value);
+        }) > 0) {
+            event(new RolesChanged($this));
+        }
+
+        return $this;
+    }
+
+    /**
+     * Revoke the given role from the model.
+     *
+     * @param string|Role $role
+     *
+     * @return self
+     */
+    public function removeRole($role): self
+    {
+        $count = $this->roles()->detach($this->getStoredRole($role));
+        $this->load('roles');
+
+        if ($this instanceof User && $count > 0) {
+            event(new RolesChanged($this));
+        }
+
+        return $this;
+    }
+
+    /**
+     * Remove all current roles and set the given ones.
+     *
+     * @param array|Role|string ...$roles
+     *
+     * @return $this
+     */
+    public function syncRoles(...$roles): self
+    {
+        $roles = $this->roleInputToIds($roles);
+
+        $changes = $this->roles()->sync($roles);
+
+        if ($this->fresh()->requiresPassword() && ! $this->hasPassword()) {
+            // Invalidate User's Session, forcing them through Auth to set a secondary password
+            $this->setRememberToken(null);
+            $this->save();
+        }
+
+        if ($this instanceof User && collect($changes)->sum(function ($value) {
+            return count($value);
+        }) > 0) {
+            event(new RolesChanged($this));
+        }
+
+        return $this;
+    }
+
+    /**
+     * Determine if the model has (one of) the given role(s).
+     *
+     * @param string|int|array|Role $roles
+     * @return bool
+     */
+    public function hasRole($roles): bool
```

Comment:
Can this be split up into more readable methods that calls this base method?

`hasRoleByName`
`hasRoleById`

Reply:
Again, is from spatie but could add new methods for that if you think its work it? I think there is merit to it being able to support all of them in one function. Admittedly it does mean lots of `if`s in the body but I don't think it looks that bad?

Reply:
I don't really mind but I think it's just a lot of logic in one function. I think it can be improved even if this one uses those new methods

Reply:
Have done some splitting

Reply:
Looks good!